flink-dstl/flink-dstl-dfs/src/test/java/org/apache/flink/changelog/fs/ChangelogStorageMetricsTest.java

```diff
@@ -295,6 +346,55 @@ public void close() {
         }
     }
 
+    private static class WaitingMaxAttemptUploader implements StateChangeUploader {
+        private final ConcurrentHashMap<UploadTask, CountDownLatch> remainingAttemptsPerTask;
+        private final int maxAttempts;
+
+        public WaitingMaxAttemptUploader(int maxAttempts) {
+            if (maxAttempts < 1) {
+                throw new IllegalArgumentException("maxAttempts < 0");
+            }
+            this.maxAttempts = maxAttempts;
+            this.remainingAttemptsPerTask = new ConcurrentHashMap<>();
+        }
+
+        @Override
+        public UploadTasksResult upload(Collection<UploadTask> tasks) throws IOException {
+
+            for (UploadTask uploadTask : tasks) {
+                CountDownLatch remainingAttempts = remainingAttemptsPerTask.get(uploadTask);
+                if (remainingAttempts == null) {
+                    remainingAttempts = new CountDownLatch(maxAttempts - 1);
+                    remainingAttemptsPerTask.put(uploadTask, remainingAttempts);
+                } else {
+                    remainingAttempts.countDown();
+                }
+            }
+            for (UploadTask uploadTask : tasks) {
+                CountDownLatch remainingAttempts = remainingAttemptsPerTask.get(uploadTask);
+                try {
+                    remainingAttempts.await();
```

Comment:
I was thinking that only the 1st attempt should be waiting, all the other should fail by throwing an exception. That would more likely catch any bugs.
WDYT?

Reply:
If last attempt throw an exception, the upload may failed, so I make first and last attempt be waiting , and others throw an exception. WDYT ?

Reply:
You're right, completing first and last attempts sounds good.